File: SerializeandDeserializeBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        s = ""
        if not root:
            s += "null,"
        else:
            s += str(root.val) + ","
            s += self.serialize(root.left)
            s += self.serialize(root.right)
        return s
    
    #this is un used.
#     def buildstring(self,node,s):
#         if not node:
#             s += "null,"
#         else:
#             s += str(node.val) + ","
#             self.buildstring(node.left,s)
#             self.buildstring(node.right,s)
#         return s
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        lst = data.split(",")
        
        return self.buildtree(lst)
        
    def buildtree(self, lst):
        s = lst.pop(0)
        if s == "null" or s == "":
            return None
        else:
            node = TreeNode(int(s))
            node.left = self.buildtree(lst)
            node.right = self.buildtree(lst)
            return node
            

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: serializeAndDeserializeBinaryTree.kt
// used the default in order traversal
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Codec() {
	// Encodes a URL to a shortened URL.
    fun serialize(root: TreeNode?): String {
        val sb = StringBuilder("")
        fun serializeHelper(node: TreeNode?) {
            if(node == null) sb.append("null,")
            else {
                sb.append(node!!.`val`.toString() + ",")
                serializeHelper(node!!.left)
                serializeHelper(node!!.right)
            }
        }
        serializeHelper(root)
        return sb.toString()
    }

    // Decodes your encoded data to tree.
    fun deserialize(data: String): TreeNode? {
        val queue = ArrayDeque(data.split(","))
        fun deserializeHelper(): TreeNode?{
            val rootStr = queue.removeFirst()
            if(rootStr == "null"){
                return null
            } else{
                val root = TreeNode(rootStr.toInt())
                root.left = deserializeHelper()
                root.right = deserializeHelper()
                return root
            }
        }
        return deserializeHelper()
    }
}

/**
 * Your Codec object will be instantiated and called as such:
 * var ser = Codec()
 * var deser = Codec()
 * var data = ser.serialize(longUrl)
 * var ans = deser.deserialize(data)
 */
